File: Data Structure/Hash/Medium/LET_no621_Task Scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        if n == 0:
            return len(tasks)
        
        memo = collections.Counter(tasks)
        max_freq = max(memo.values())
        res = (max_freq - 1) * (n + 1)
        
        for f in memo.values():
            if f == max_freq:
                res += 1
        
        return max(len(tasks), res)

        """
        example 1: memo = {'A':3, 'B':3}, n = 2
        prepare: A X X A X X 
        roop 1: A X X A X X A
        roop 2: A B X A B X A B

        example 2: memo = {'A':6, 'B':1, 'C':1, 'D':1, 'E':1, 'F':1, 'G':1}, n = 2 
        prepare: A X X A X X A X X A X X A X X
        roop 1: A X X A X X A X X A X X A X X A
        roop 2 - 7: A B C A D E A F G A X X A X X A
        """ 

File: Data Structure/Hash/Medium/LET_no621_Task Scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> memo;
        int maxFreq = 0;
        for (char c : tasks) {
            memo[c]++;
            maxFreq = max(maxFreq, memo[c]);
        }

        int res = (maxFreq - 1) * (n + 1);
        for (auto it = memo.begin(); it != memo.end(); it++)
            res += it->second == maxFreq;

        return max(int(tasks.size()), res);
    }
};